kubernetes/scheduler/controller.go

```diff
@@ -0,0 +1,668 @@
+
+	return controller
+}
+
+// Run sets up event handlers, synces informer caches and starts workers. It
+// blocks until stopCh is closed, at which point it'll shut down workqueue and
+// wait for workers to finish processing.
+func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
+	c.Logger = c.opts.Logger.New("schedulerController", "kubernetes")
+	kubeLogFile := filepath.Join(c.opts.ManagerDir, kubeSchedulerControllerLog)
+	fh, err := log15.FileHandler(kubeLogFile, log15.LogfmtFormat())
+	if err != nil {
+		return fmt.Errorf("wr kubernetes scheduler could not log to %s: %s", kubeLogFile, err)
+	}
+
+	l15h.AddHandler(c.Logger, fh)
+	c.Debug("In Run()")
+	defer runtime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	// Start informer factories, begin populating informer caches.
+
+	// Wait for caches to sync before starting workers
+	c.Debug("Waiting for caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.podSynced, c.nodeSynced); !ok {
+		c.Crit("failed to wait for caches to sync")
+		utilruntime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+	c.Debug("Caches synced")
+
+	c.Debug("In Run(), starting workers")
+	// start workers
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runReqCheck, time.Second, stopCh)
+		go wait.Until(c.runPodAlive, time.Second, stopCh)
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+
+	return nil
+}
+
+// continually call processNextWorkItem to read and process the next message on
+// the workqueue
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem reads a single work item off the workqueue and attempts
+// to process it, by calling the syncHandler.
+func (c *Controller) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	// Wrap this block in func so can defer c.workqueue.Done.
+	err := func(obj interface{}) error {
+		// Call done so workqueue knows we have finished processing this item
+		// Must also call forget if we don't want the item re-queued.
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		// strings come off workqueue: namespace/name. delayed nature of wq
+		// means items in informer cache may be more up to date than the item
+		// initially put in the wq.
+		if key, ok = obj.(string); !ok {
+			// As item in workqueue is invalid, call forget else would loop
+			// attempting to process an invalid work item.
+			c.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+
+		// Run syncHandler, passing it the namespace/name key of the resource to
+		// be synced
+		err := c.processItem(key)
+		// handleErr will handle adding to the queue again.
+		c.handleErr(err, key)
+		// Finally, if no error occurs forget the item so it's not queued again
+		c.workqueue.Forget(obj)
```

Comment:
You do this even if an error occurs. Intended?

Reply:
Yep, It's recommended best practice for controllers / operators in k8s world. When I  prodded the slack channel I was redirected here: https://en.wikipedia.org/wiki/Optimistic_concurrency_control

```go
	// maxRetries is the number of times a deployment will be retried before it
	// is dropped out of the queue. With the current rate-limiter in use
	// (5ms*2^(maxRetries-1)) the following numbers represent the times a work
	// item is going to be requeued:
	//
	// 5ms, 10ms, 20ms, 40ms, 80ms, 160ms, 320ms, 640ms, 1.3s, 2.6s, 5.1s,
	// 10.2s, 20.4s, 41s, 82s
	maxRetries = 15
```

Defines how many times an item is added to the queue again. The idea here is that we aren't the only actor on the resources we get in the queue, something else may compete for it (or change it without us being able to do anything about it) so the result is a mechanism that waits a little while, and then tries again hoping for the best.

Reply:
https://trello.com/c/YPBw53V0/43-writing-kubernetes-controllers 

Has lots of reading in it.